### PR TITLE
CORDA-3464: Also scan attachment:// URLs for custom serializers.

### PR DESCRIPTION
**Pull request by chrisr3**
Ensure that we scan attachments for instances of custom serializers.

### CODE REVIEW

File: core/src/test/kotlin/net/corda/core/internal/URLSchemes.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package net.corda.core.internal
+
+import net.corda.core.serialization.internal.AttachmentURLStreamHandlerFactory
+import java.net.URL
+
+/**
+ * A JVM can only ever invoke [URL.setURLStreamHandlerFactory] once.
+ */
+object URLSchemes {
```
Comment (mnesbit) [inline core/src/test/kotlin/net/corda/core/internal/URLSchemes.kt]: Where in our init sequence is this pulled in?

Comment (chrisr3) [reply core/src/test/kotlin/net/corda/core/internal/URLSchemes.kt]: This is a test class. It does Testy Things testily in the Land of Test.

**Review by tudor-malene — COMMENTED** [review]
I think it looks good. 
But unfortunately, due to how complex everything is, this test doesn't prove that it actually works how we expect.
To really test the functionality, a complex driver test with out of process nodes that does this:

1. defines a custom serializer in a flow. This should no longer break transaction verification.
2. defines a custom serializer in a state and actually uses it.
3. Send this transaction to a node that doesn't have that cordapp installed, and it should still work


(But I guess this can also be tested manually with the project that was created by the reporter)

**Comment by chrisr3** [comment]
> But unfortunately, due to how complex everything is, this test doesn't prove that it actually works how we expect.

Such a complex test would belong in `core-tests`. The current unit test is _literally_ just for `ClassLoadingUtils`.

**Comment by tudor-malene** [comment]
> > But unfortunately, due to how complex everything is, this test doesn't prove that it actually works how we expect.
> 
> Such a complex test would belong in `core-tests`. The current unit test is _literally_ just for `ClassLoadingUtils`.

Totally. 
It's just that I can't asses based on it if this fix actually behaves as we'd expect.
But it is definitely a valuable test.

File: node/src/integration-test/kotlin/net/corda/contracts/serialization/custom/CustomSerializerContract.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    override fun verify(tx: LedgerTransaction) {
+        val currancyData = tx.outputsOfType(CurrancyState::class.java)
+        require(currancyData.isNotEmpty()) {
+            "Requires at least one currancy state"
+        }
+
+        currancyData.forEach {
+            require(it.currancy.currants in 0..MAX_CURRANT) {
+                "Too many currants! ${it.currancy.currants} is unraisinable!"
```
Comment (rick-r3) [inline node/src/integration-test/kotlin/net/corda/contracts/serialization/custom/CustomSerializerContract.kt]: Very fine work

File: node/src/integration-test/kotlin/net/corda/contracts/serialization/custom/CurrancySerializer.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+@Suppress("unused")
+class CurrancySerializer : SerializationCustomSerializer<Currancy, CurrancySerializer.Proxy> {
+    data class Proxy(val currants: Long)
+
+    override fun fromProxy(proxy: Proxy): Currancy = Currancy(proxy.currants)
+    override fun toProxy(obj: Currancy) = Proxy(obj.currants)
+}
+
+data class Currancy(val currants: Long) : Comparable<Currancy> {
```
Comment (rick-r3) [inline node/src/integration-test/kotlin/net/corda/contracts/serialization/custom/CurrancySerializer.kt]: Maybe `Currantsy` ?  The temptation to correct the spelling is very strong.  I almost commented before I realised it was intentional!

**Review by rick-r3 — DISMISSED** [review]

**Review by tudor-malene — APPROVED** [review]
LGTM!

(Maybe worth raising a story to add more tests in the future like the ones I've mentioned earlier).

**Pull request merged**